File: android/app/src/main/java/ru/techgid/domain/model/Car.kt
package ru.techgid.domain.model

/**
 * Domain-модели автомобилей.
 */
data class CarBrand(
    val id: Int,
    val name: String,
    val slug: String,
    val logoUrl: String? = null,
    val country: String? = null,
)

data class CarModel(
    val id: Int,
    val brandId: Int,
    val name: String,
    val slug: String,
)

data class CarGeneration(
    val id: Int,
    val modelId: Int,
    val name: String,
    val slug: String,
    val chassisCode: String? = null,
    val yearStart: Int,
    val yearEnd: Int? = null,
    val imageUrl: String? = null,
)

data class CarEngine(
    val id: Int,
    val generationId: Int,
    val code: String,
    val name: String,
    val displacementLabel: String? = null,
    val fuelType: String,
    val powerHp: Int? = null,
)

data class CarConfiguration(
    val id: Int,
    val generationId: Int,
    val engineId: Int,
    val displayName: String? = null,
)

/**
 * Полный выбор пользователя — от марки до конфигурации.
 */
data class CarSelection(
    val brand: CarBrand? = null,
    val model: CarModel? = null,
    val generation: CarGeneration? = null,
    val engine: CarEngine? = null,
    val configuration: CarConfiguration? = null,
) {
    val isComplete: Boolean
        get() = brand != null && model != null && generation != null && engine != null

    val displayTitle: String
        get() = buildString {
            brand?.let { append(it.name) }
            model?.let { append(" ${it.name}") }
            generation?.let { append(" (${it.yearStart})") }
            engine?.let { append(" ${it.displacementLabel ?: it.name}") }
        }.trim()
}

File: backend/app/schemas/car.py
"""Pydantic-схемы для автомобилей."""

from pydantic import BaseModel, ConfigDict

from app.models.car import BodyType, DriveType, FuelType, TransmissionType


# ── Марка ───────────────────────────────────────────────────────

class CarBrandOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    slug: str
    logo_url: str | None = None
    country: str | None = None


class CarBrandCreate(BaseModel):
    name: str
    slug: str
    logo_url: str | None = None
    country: str | None = None
    sort_order: int = 0


# ── Модель ──────────────────────────────────────────────────────

class CarModelOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    brand_id: int
    name: str
    slug: str


class CarModelCreate(BaseModel):
    brand_id: int
    name: str
    slug: str
    sort_order: int = 0


# ── Поколение ───────────────────────────────────────────────────

class CarGenerationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    model_id: int
    name: str
    slug: str
    chassis_code: str | None = None
    year_start: int
    year_end: int | None = None
    image_url: str | None = None


class CarGenerationCreate(BaseModel):
    model_id: int
    name: str
    slug: str
    chassis_code: str | None = None
    year_start: int
    year_end: int | None = None
    image_url: str | None = None


# ── Двигатель ───────────────────────────────────────────────────

class CarEngineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    generation_id: int
    code: str
    name: str
    displacement_label: str | None = None
    fuel_type: FuelType
    power_hp: int | None = None
    torque_nm: int | None = None


class CarEngineCreate(BaseModel):
    generation_id: int
    code: str
    name: str
    displacement_cc: int | None = None
    displacement_label: str | None = None
    fuel_type: FuelType
    power_hp: int | None = None
    torque_nm: int | None = None


# ── Кузов ───────────────────────────────────────────────────────

class CarBodyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    generation_id: int
    body_type: BodyType
    doors_count: int
    name: str | None = None


# ── Комплектация ────────────────────────────────────────────────

class CarTrimOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    generation_id: int
    name: str
    market: str | None = None


# ── Конфигурация ────────────────────────────────────────────────

class CarConfigurationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    generation_id: int
    engine_id: int
    body_id: int | None = None
    trim_id: int | None = None
    transmission_type: TransmissionType | None = None
    drive_type: DriveType | None = None
    year_start: int | None = None
    year_end: int | None = None
    display_name: str | None = None


class CarConfigurationCreate(BaseModel):
    generation_id: int
    engine_id: int
    body_id: int | None = None
    trim_id: int | None = None
    transmission_type: TransmissionType | None = None
    drive_type: DriveType | None = None
    year_start: int | None = None
    year_end: int | None = None
    display_name: str | None = None


# ── Полная иерархия для выбора авто ─────────────────────────────

class CarBrandWithModels(CarBrandOut):
    models: list[CarModelOut] = []


class CarModelWithGenerations(CarModelOut):
    generations: list[CarGenerationOut] = []


class CarGenerationFull(CarGenerationOut):
    engines: list[CarEngineOut] = []
    bodies: list[CarBodyOut] = []
    trims: list[CarTrimOut] = []
